frontend/src/utilities/notebookControllerUtils.ts

```diff
@@ -30,13 +30,21 @@ import {
 } from '../types';
 import AppContext from '../app/AppContext';
 
-export const usernameTranslate = (username: string): string =>
-  username
+export const usernameTranslate = (username: string): string => {
```

Comment:
Write jest/unit tests for this function. This is going to need to be pure on characters we don't usually see. So we cannot be just hoping it works.

Reply:
Since we are doing manual translation of characters to a well defined standard, Is there a hardened library that already handles this usecase?

Reply:
🤔  Hmmm -- not from a quick google. The logic needs to be tested to sorta meet something like `^[a-z0-9-]{1,63}$` or maybe even more specific `^[a-z]{1}[a-z0-9-]{1,62}$` so those that need to always start with alphabetical characters work too. The function is less about usernames and more about taking a string that is viable to work with k8s strict naming.

I don't think it's that crazy to write tests to verify you have broad strokes covered. It's a pure functional utility, it should be easy to test a bunch of scenarios.

Reply:
I agree completely on the test coverage.  I didn't know if there was a library available like `encodeURIComponent()` that we could use instead of the chained `.replace()` calls

Reply:
OH haha... I thought you were questioning the custom test approach lol. My apologizes. There is no built in JS function for K8s names... only dealing with the URL. So we need to definitely approach what is it for K8s that we need to do.

Reply:
> Write jest/unit tests for this function. This is going to need to be pure on characters we don't usually see. So we cannot be just hoping it works.

I'll add some unit tests for the usernames that contain special characters. Regarding the comments here and in #410, should the constraints (username length and that the username has to start (and also end) with an alphanumeric character) be addressed somewhere?

Reply:
I think today the function is only for users -- and we prefix it always... both of which prevent any starting/ending issues I think. Length is perhaps a concern, but I am not sure how we can do that right now deterministically, so lets not worry about it yet. I think 63 characters is a safe window to manage that problem without actually managing it.

Might be good to handle the edge case but let us not slow down this work on that. (my opinion -- @LaVLaS thoughts?)

Reply:
I agree with the character limit that aligns with the k8s char limit